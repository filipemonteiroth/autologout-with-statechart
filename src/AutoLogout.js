import React from "react";
import { withStatechart, State } from "react-automata";

export const statechart = {
  initial: "loggedIn",
  states: {
    loggedIn: {
      onEntry: "addMouseMoveListener",
      onExit: "removeMouseMoveListener",
      on: {
        LOGOUT: "loggedOff"
      },
      initial: "working",
      states: {
        working: {
          onEntry: ["resetTimers", "startIdleTimer"],
          on: {
            MOUSE_MOVE: "working",
            IDLE: "idle"
          }
        },
        idle: {
          onEntry: "startLogoutTimer",
          onExit: "resetTimers",
          on: {
            MOUSE_MOVE: "working"
          }
        }
      }
    },
    loggedOff: {
      on: {
        LOGGED_IN: "loggedIn"
      }
    }
  }
};

class AutoLogout extends React.Component {
  idleTimer = null;
  logoutTimer = null;

  mousemoveListener = () => {
    this.props.transition("MOUSE_MOVE");
  };

  addMouseMoveListener = () => {
    document.addEventListener("mousemove", this.mousemoveListener);
  };

  removeMouseMoveListener = () => {
    document.removeEventListener("mousemove", this.mousemoveListener);
  };

  resetTimers = () => {
    clearTimeout(this.idleTimer);
    clearTimeout(this.logoutTimer);
  };

  startIdleTimer = () => {
    const { transition } = this.props;
    this.idleTimer = setTimeout(() => transition("IDLE"), 5000);
  };

  startLogoutTimer = () => {
    const { transition } = this.props;
    this.logoutTimer = setTimeout(() => transition("LOGOUT"), 5000);
  };

  render() {
    const { children } = this.props;
    return (
      <div style={containerStyle}>
        <div style={styles}>
          <State value="loggedIn.working">You are working</State>
          <State value="loggedIn.idle">You are about to be logged out</State>
        </div>
        <State value="loggedOff">
          <div>
            <p>You are logged off</p>
            <button onClick={() => this.props.transition("LOGGED_IN")}>
              Login
            </button>
          </div>
        </State>
        <State value="loggedIn.*">{children}</State>
      </div>
    );
  }
}

const containerStyle = {
  backgroundColor: "#f1f1f1",
  height: "100%",
  width: "100%"
};

const styles = {
  backgroundColor: "#0053ad",
  height: "40px",
  color: "#ffffff",
  fontSize: "22px",
  display: "flex",
  justifyContent: "center",
  alignItems: "center"
};

export { AutoLogout };

export default withStatechart(statechart, { devTools: true })(AutoLogout);
