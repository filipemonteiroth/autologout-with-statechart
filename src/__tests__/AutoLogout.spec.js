import { testStatechart } from "react-automata";

import { AutoLogout, statechart } from "../AutoLogout";

it("tests statechart automatically", () => {
  testStatechart({ statechart }, AutoLogout);
});
