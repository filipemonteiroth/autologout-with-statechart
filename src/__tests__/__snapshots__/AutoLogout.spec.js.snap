// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loggedIn.idle 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#f1f1f1",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0053ad",
        "color": "#ffffff",
        "display": "flex",
        "fontSize": "22px",
        "height": "40px",
        "justifyContent": "center",
      }
    }
  >
    You are about to be logged out
  </div>
</div>
`;

exports[`loggedIn.working 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#f1f1f1",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0053ad",
        "color": "#ffffff",
        "display": "flex",
        "fontSize": "22px",
        "height": "40px",
        "justifyContent": "center",
      }
    }
  >
    You are working
  </div>
</div>
`;

exports[`loggedOff 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#f1f1f1",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0053ad",
        "color": "#ffffff",
        "display": "flex",
        "fontSize": "22px",
        "height": "40px",
        "justifyContent": "center",
      }
    }
  />
  <div>
    <p>
      You are logged off
    </p>
    <button
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
